Settings file creation and validation (#2176)
PowerShell/PSScriptAnalyzer: docs/Cmdlets/New-ScriptAnalyzerSettingsFile.md

---
external help file: Microsoft.Windows.PowerShell.ScriptAnalyzer.dll-Help.xml
Module Name: PSScriptAnalyzer
ms.date: 04/17/2026
schema: 2.0.0
---

# New-ScriptAnalyzerSettingsFile

## SYNOPSIS
Creates a new PSScriptAnalyzer settings file.

## SYNTAX

```
New-ScriptAnalyzerSettingsFile [[-Path] <string>] [-BaseOnPreset <string>] [-Force] [-WhatIf] [-Confirm] [<CommonParameters>]
```

## DESCRIPTION

The `New-ScriptAnalyzerSettingsFile` cmdlet creates a `PSScriptAnalyzerSettings.psd1` file in the
specified directory.

When the **BaseOnPreset** parameter is provided, the generated file contains the rules and
configuration defined by the given preset.

When **BaseOnPreset** is not provided, the generated file includes all current rules in the
`IncludeRules` list and populates the `Rules` section with all configurable properties, set to their
default values. Both modes also include `CustomRulePath`, `RecurseCustomRulePath`, and
`IncludeDefaultRules` keys with descriptive comments so the file is immediately ready for

> customisation.

Reviewer: Our style guide dictates American spellings.

```suggestion
customization.
```

If a settings file already exists at the target path, the cmdlet emits a terminating error unless

> the **Force** parameter is specified - in which case it is overwritten.

Reviewer: ```suggestion
the **Force** parameter is specified - in which case it's overwritten.
```

Parameter Sets: (All)
Aliases:

Required: False
Position: Named
Default value: False
Accept pipeline input: False
Accept wildcard characters: False
```

### -Confirm

Prompts you for confirmation before running the cmdlet.

```yaml
Type: SwitchParameter
Parameter Sets: (All)
Aliases: cf

Required: False
Position: Named
Default value: False
Accept pipeline input: False
Accept wildcard characters: False
```

### CommonParameters

This cmdlet supports the common parameters: -Debug, -ErrorAction, -ErrorVariable,
-InformationAction, -InformationVariable, -OutVariable, -OutBuffer, -PipelineVariable, -Verbose,
-WarningAction, and -WarningVariable. For more information, see
[about_CommonParameters](http://go.microsoft.com/fwlink/?LinkID=113216).

## INPUTS

### None

## OUTPUTS

### System.IO.FileInfo

The cmdlet returns a **FileInfo** object representing the created settings file.

## NOTES

The output file is always named `PSScriptAnalyzerSettings.psd1` so that the automatic settings
discovery in `Invoke-ScriptAnalyzer` picks it up when analysing scripts in the same directory.

Note: Relative paths in `CustomRulePath` are resolved from the caller's current working directory,

> not from the location of the settings file. This matches `Invoke-ScriptAnalyzer` behaviour.

Reviewer: ```suggestion
Relative paths in `CustomRulePath` are resolved from the caller's current working directory,
not from the location of the settings file. This matches `Invoke-ScriptAnalyzer` behavior.
```
